### ISSUE TITLE
Stabilize the operation if multicast traffic isn't available or lossy

### ISSUE THREAD

**Issue by syssi** [Danielhiversen/PyXiaomiGateway#26]
1. Check if multicast messages was received in the past ("iam" discovery response / heartbeats) and log a warning.
2. Recovery from "Invalid key" by retrieval of a new token (get_id_list)
3. If every executed command needs a "token recovery" report a degraded performance because of the lack of multicast traffic.
4. Implement an indicator if a command execution was successful (without invalid key response) to be sure a valid encryption key is used.

cp. https://community.home-assistant.io/t/pyxiaomigateway-problem-error-invalid-key/28168
https://community.home-assistant.io/t/xiaomi-gateway-integration-help-needed/28563/10
https://community.home-assistant.io/t/new-xiaomi-gateway-firmware-xiaomi-aqara-component/27637/30
https://community.home-assistant.io/t/xiaomi-gateway-not-discovered/25030/158
https://community.home-assistant.io/t/xiaomi-sensors-not-working-after-0-57-update-solved/31693/6
https://community.home-assistant.io/t/xiaomi-gateway-is-recognized-but-does-not-take-any-commands-or-brings-data-help-needed/28725

This is my explanation of the "invalid key" error: 

Some responses (get_id_list (direct message), heartbeat (multicast)) of the gateway contains a "token" which is valid for a short period of time. The combination of the token plus the encryption key of the android app is used to calculate the "key" value of a "write" command:

```
{"cmd": "write", "sid": "34ce00909ed1", "data": {"rgb": 3036676095, "key": "5059aeccf37f66b5cfec0cdfaab76e29"}}
```

I don't know the behavior of the gateway exactly but if you miss one (or multiple) of the temporary tokens a old/invalid token is used and the misleading "Invalid key" error pops up although your encryption key (android key) is fine.

We will implement a safeguard for the "Invalid key" error so a new token will be requested and a retry of the failed command wil be sent. Do not forget: The cause of the issue is packet loss. :-(

cp. https://github.com/Danielhiversen/PyXiaomiGateway/issues/21

**Comment by Danielhiversen** [comment]
I have started working on it here: https://github.com/Danielhiversen/PyXiaomiGateway/compare/invalid_key?expand=1
But do not have time to finish it now.

**Comment by syssi** [comment]
I will use your code as basis and try to provide some iptable rules for testing.

**Comment by TSM-EVO** [comment]
Is the resolution to this going to mean no reliance on multicast?

There's a history of multicast packets not functioning in Home Assistant that means lots of users with these gateways end up with timeouts after 2.5 hours. 

For docker users who can't utilize multicast packets due to using a reverse proxy such as through traefik, this also means that devices stop working after that time.

In case you're curious: https://community.home-assistant.io/t/xiaomi-gateway-becomes-unavailable-after-some-hours/50154/34

**Comment by PaulAnnekov** [comment]
> Is the resolution to this going to mean no reliance on multicast?

No, it's just fixing some minor issues with communication between client and hub.

> There's a history of multicast packets not functioning in Home Assistant that means lots of users with these gateways end up with timeouts after 2.5 hours.

That's because their network is misconfigured. Aqara hub doesn't provide another pushing communication method, except multicasting.

**Comment by syssi** [comment]
Duplicates and not chronological traffic should be handled, too.

**Comment by andriej** [comment]
Anything done with this issue?

**Comment by andriej** [comment]
@syssi any way this can be fixed? currently I have issue where after HA restart all 3 gateways are unavailable even if I see MCAST traffic in debug log

**Comment by syssi** [comment]
@andriej Please open a new issue. It looks like you've another (network!) problem.

**Comment by andriej** [comment]
@syssi - opened, with log and description
